compiler/rustc_lint/src/non_fmt_panic.rs

```diff
@@ -0,0 +1,197 @@
+use crate::{LateContext, LateLintPass, LintContext};
+use rustc_ast as ast;
+use rustc_errors::{pluralize, Applicability};
+use rustc_hir as hir;
+use rustc_middle::ty;
+use rustc_parse_format::{ParseMode, Parser, Piece};
+use rustc_span::{sym, symbol::kw, InnerSpan, Span, Symbol};
+
+declare_lint! {
+    /// The `non_fmt_panic` lint detects `panic!(..)` invocations where the first
+    /// argument is not a formatting string.
+    ///
+    /// ### Example
+    ///
+    /// ```rust,no_run
+    /// panic!("{}");
+    /// panic!(123);
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// In Rust 2018 and earlier, `panic!(x)` directly uses `x` as the message.
+    /// That means that `panic!("{}")` panics with the message `"{}"` instead
+    /// of using it as a formatting string, and `panic!(123)` will panic with
+    /// an `i32` as message.
+    ///
+    /// Rust 2021 always interprets the first argument as format string.
+    NON_FMT_PANIC,
+    Warn,
+    "detect single-argument panic!() invocations in which the argument is not a format string",
+    report_in_external_macro
+}
+
+declare_lint_pass!(NonPanicFmt => [NON_FMT_PANIC]);
+
+impl<'tcx> LateLintPass<'tcx> for NonPanicFmt {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx hir::Expr<'tcx>) {
+        if let hir::ExprKind::Call(f, [arg]) = &expr.kind {
+            if let &ty::FnDef(def_id, _) = cx.typeck_results().expr_ty(f).kind() {
+                if Some(def_id) == cx.tcx.lang_items().begin_panic_fn()
+                    || Some(def_id) == cx.tcx.lang_items().panic_fn()
+                    || Some(def_id) == cx.tcx.lang_items().panic_str()
+                {
+                    if let Some(id) = f.span.ctxt().outer_expn_data().macro_def_id {
+                        if cx.tcx.is_diagnostic_item(sym::std_panic_2015_macro, id)
+                            || cx.tcx.is_diagnostic_item(sym::core_panic_2015_macro, id)
+                        {
+                            check_panic(cx, f, arg);
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
+
+fn check_panic<'tcx>(cx: &LateContext<'tcx>, f: &'tcx hir::Expr<'tcx>, arg: &'tcx hir::Expr<'tcx>) {
+    if let hir::ExprKind::Lit(lit) = &arg.kind {
+        if let ast::LitKind::Str(sym, _) = lit.node {
+            // The argument is a string literal.
+            check_panic_str(cx, f, arg, &sym.as_str());
+            return;
+        }
+    }
+
+    // The argument is *not* a string literal.
+
+    let (span, panic) = panic_call(cx, f);
+
+    cx.struct_span_lint(NON_FMT_PANIC, arg.span, |lint| {
+        let mut l = lint.build("panic message is not a string literal");
+        l.note("this is no longer accepted in Rust 2021");
```

Comment:
I think we always mention the "edition" when talking about them: "Rust 2021 edition"

Reply:
There used to be a mix of "Rust 20xx" and "edition 20xx". As part of https://github.com/rust-lang/rust/pull/79576 this was changed to consistently use "Rust 20xx" everywhere: https://github.com/rust-lang/rust/pull/79576/commits/c574ded57da4c7c631b836c397f31d2a72dc1280

E.g.
https://github.com/rust-lang/rust/blob/d4e3570db4c007089035b833cc20c7fc2f8cb32f/compiler/rustc_parse/src/parser/item.rs#L1678-L1679